### PR TITLE
Extensions to support input collection initializers with union types

### PR DESCRIPTION
**Pull request by mikhailshilkov**
Extension methods to support input collection initializers with union types

Resolve https://github.com/pulumi/pulumi-azure-nextgen/issues/173

### CODE REVIEW

File: sdk/dotnet/Pulumi/Core/Input.cs [hunk 1]
```diff
@@ -75,5 +75,25 @@ public static class InputExtensions
...
+        
+        public static void Add<T, U>(this InputList<Union<T, U>> map, Input<T> value)
+        {
+            map.Add(value.ToOutput().Apply(v => (Union<T, U>)v));
+        }
+        
+        public static void Add<T, U>(this InputList<Union<T, U>> map, Input<U> value)
+        {
+            map.Add(value.ToOutput().Apply(v => (Union<T, U>)v));
+        }
```
Comment (justinvp) [inline sdk/dotnet/Pulumi/Core/Input.cs:97]: I could also imagine creating separate `InputMapExtensions` and `InputListExtensions` classes, but don't see a ton of value/need to do that, so I'm fine with them being on `InputExtensions`.

Comment (justinvp) [inline sdk/dotnet/Pulumi/Core/Input.cs:97]:
```suggestion
        public static void Add<T, U>(this InputList<Union<T, U>> list, Input<T> value)
        {
            list.Add(value.ToOutput().Apply(v => (Union<T, U>)v));
        }
        
        public static void Add<T, U>(this InputList<Union<T, U>> list, Input<U> value)
        {
            list.Add(value.ToOutput().Apply(v => (Union<T, U>)v));
        }
```

File: sdk/dotnet/Pulumi/PublicAPI.Shipped.txt [hunk 2]
```diff
@@ -245,6 +245,10 @@ static Pulumi.Deployment.TestWithServiceProviderAsync<TStack>(Pulumi.Testing.IMo
 static Pulumi.Input<T>.implicit operator Pulumi.Input<T>(Pulumi.Output<T> value) -> Pulumi.Input<T>
 static Pulumi.Input<T>.implicit operator Pulumi.Input<T>(T value) -> Pulumi.Input<T>
 static Pulumi.Input<T>.implicit operator Pulumi.Output<T>(Pulumi.Input<T> input) -> Pulumi.Output<T>
+static Pulumi.InputExtensions.Add<T, U>(this Pulumi.InputList<Pulumi.Union<T, U>> map, Pulumi.Input<T> value) -> void
+static Pulumi.InputExtensions.Add<T, U>(this Pulumi.InputList<Pulumi.Union<T, U>> map, Pulumi.Input<U> value) -> void
```
Comment (justinvp) [inline sdk/dotnet/Pulumi/PublicAPI.Shipped.txt:249]:
```suggestion
static Pulumi.InputExtensions.Add<T, U>(this Pulumi.InputList<Pulumi.Union<T, U>> list, Pulumi.Input<T> value) -> void
static Pulumi.InputExtensions.Add<T, U>(this Pulumi.InputList<Pulumi.Union<T, U>> list, Pulumi.Input<U> value) -> void
```

**Review by justinvp — APPROVED** [review]
LGTM otherwise

File: sdk/dotnet/Pulumi/Core/Input.cs (hunk 1 above)
Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Core/Input.cs:97]: Yeah, I was on a fence there. Do you know of any best practices/conventions around that?

Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Core/Input.cs:97]: 🤦 thank you!

Comment (mikhailshilkov) [reply sdk/dotnet/Pulumi/Core/Input.cs]: After reading some more, I went ahead and split it.

**Pull request merged**